### PR TITLE
chore(gen2-storage): Update StorageImage API

### PR DESCRIPTION
**Pull request by esauerbo**
#### Checklist

- [ ] Have read the [Pull Request Guidelines](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md)
- [ ] PR description included
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] `yarn test` passes and tests are updated/added
- [ ] No side effects or [`sideEffects`](https://github.com/aws-amplify/amplify-ui/blob/main/packages/react/CONTRIBUTING.md#code-standards) field updated

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/react/src/hooks/useGetURL.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        if (ignore) {
+          return;
+        }
+        if (isFunction(onError)) {
+          onError(error);
+        }
+
+        return () => {
+          ignore = true;
+        };
```
Comment (calebpollman) [inline packages/react/src/hooks/useGetURL.ts:70]: Why is this here?

File: packages/react/src/hooks/useGetURL.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+): string | (({ identityId }: { identityId?: string }) => string) {
+  if (input.key !== undefined) {
+    return input.key;
+  }
+  return input.path;
+}
+
+type UseGetUrl = (input: UseGetUrlInput) => UseGetUrlOutput;
+
+export const useGetURL: UseGetUrl = (
```
Comment (calebpollman) [inline packages/react/src/hooks/useGetURL.ts:33]: The name of the hook (and all related filenames) should be `useGetUrl`

File: packages/react/src/hooks/useGetURL.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+import { useHasValueUpdated } from '@aws-amplify/ui-react-core';
+
+type UseGetUrlInput = GetUrlInput & {
+  onError?: (error: Error) => void;
+};
+
+interface UseGetUrlOutput {
+  isLoading: boolean;
+  url?: URL;
+  expiresAt?: Date;
```
Comment (calebpollman) [inline packages/react/src/hooks/useGetURL.ts:15]: Why are these optional?

File: packages/react/src/hooks/useGetURL.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+// interface UseGetUrlOutput extends GetUrlOutput{
+//     isLoading: boolean;
+// }
```
Comment (calebpollman) [inline packages/react/src/hooks/useGetURL.ts:20]: Please remove commented code 😅

File: packages/react/src/hooks/useGetURL.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  React.useEffect(() => {
+    if (!hasImgUpdated) {
+      return;
+    }
+
+    setIsLoading(true);
+
+    let ignore = false;
+
+    // @ts-ignore TODO why is this having an issue
```
Comment (calebpollman) [inline packages/react/src/hooks/useGetURL.ts:51]: In the future, please try to resolve TODO questions before review

File: packages/react/src/hooks/useGetURL.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+type UseGetUrl = (input: UseGetUrlInput) => UseGetUrlOutput;
+
+export const useGetURL: UseGetUrl = (
+  input: UseGetUrlInput
+): UseGetUrlOutput => {
+  const [result, setResult] = React.useState<GetUrlOutput>();
+  const [isLoading, setIsLoading] = React.useState<boolean>(false);
+  const imgLocation = getKeyOrPath(input);
+  const hasImgUpdated = useHasValueUpdated(imgLocation);
+  const { onError, ...getUrlInput } = input;
```
Comment (calebpollman) [inline packages/react/src/hooks/useGetURL.ts:40]: `getUrlInput` will have a new object reference on each call of this hook causing the `useEffect` on line 42 to run on every render. Anything we can do to prevent that?

File: packages/react/src/hooks/useGetURL.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  return input.path;
+}
+
+type UseGetUrl = (input: UseGetUrlInput) => UseGetUrlOutput;
+
+export const useGetURL: UseGetUrl = (
+  input: UseGetUrlInput
+): UseGetUrlOutput => {
+  const [result, setResult] = React.useState<GetUrlOutput>();
+  const [isLoading, setIsLoading] = React.useState<boolean>(false);
```
Comment (calebpollman) [inline packages/react/src/hooks/useGetURL.ts:37]: Only need one `useState` here

File: packages/react/src/hooks/useGetURL.ts [hunk 8]
```diff
@@ -0,0 +1,77 @@
+import * as React from 'react';
+
+import { getUrl, GetUrlInput, GetUrlOutput } from 'aws-amplify/storage';
+
+import { isFunction } from '@aws-amplify/ui';
+import { useHasValueUpdated } from '@aws-amplify/ui-react-core';
+
+type UseGetUrlInput = GetUrlInput & {
```
Comment (calebpollman) [inline packages/react/src/hooks/useGetURL.ts:8]: Prefer `interface` over `type` where possible

File: packages/react/src/hooks/useGetURL.ts [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+type UseGetUrl = (input: UseGetUrlInput) => UseGetUrlOutput;
+
+export const useGetURL: UseGetUrl = (
+  input: UseGetUrlInput
+): UseGetUrlOutput => {
+  const [result, setResult] = React.useState<GetUrlOutput>();
+  const [isLoading, setIsLoading] = React.useState<boolean>(false);
+  const imgLocation = getKeyOrPath(input);
```
Comment (calebpollman) [inline packages/react/src/hooks/useGetURL.ts:38]: Nit: `imgLocation` -> `keyOrPath`

File: packages/react/src/hooks/__tests__/useGetURL.test.ts [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+import { useGetURL } from '../useGetURL';
+
+jest.mock('aws-amplify/storage');
+
+type UseGetUrlInput = GetUrlInput & {
+  onError?: (error: Error) => void;
+};
+
+describe('useGetURL', () => {
+  afterEach(() => jest.clearAllMocks());
```
Comment (calebpollman) [inline packages/react/src/hooks/__tests__/useGetURL.test.ts:14]: Don't think `afterEach` does anything with the return value of it's callback param or that `jest.clearAllMocks` returns anything?

File: packages/react/src/hooks/__tests__/useGetURL.test.ts [hunk 11]
```diff
@@ -0,0 +1,90 @@
+import { renderHook } from '@testing-library/react-hooks';
+import { waitFor } from '@testing-library/react';
+import { getUrl, GetUrlInput } from 'aws-amplify/storage';
+
+import { useGetURL } from '../useGetURL';
+
+jest.mock('aws-amplify/storage');
+
+type UseGetUrlInput = GetUrlInput & {
```
Comment (calebpollman) [inline packages/react/src/hooks/__tests__/useGetURL.test.ts:9]: Should be imported from `../useGetUrl`

**Review by calebpollman — CHANGES_REQUESTED** [review]
Great start! To speed up the review cycle, let's break this PR down in to smaller easily digestible PRs. Please open a PR with just `useGetUrl` and its unit tests, will make it easier to address the required changes on the hook without having to update the rest of the code in the PR. We can follow up with the PR integrating `useGetUrl` in to `StorageImage`

File: packages/react/src/hooks/useGetURL.ts (hunk 3 above)
Comment (esauerbo) [reply packages/react/src/hooks/useGetURL.ts:15]: Because they are initially undefined

File: packages/react/src/hooks/useGetURL.ts (hunk 8 above)
Comment (esauerbo) [reply packages/react/src/hooks/useGetURL.ts]: I had tried that initially but get this ts error `An interface can only extend an object type or intersection of object types with statically known members.`

File: packages/react/src/hooks/__tests__/useGetURL.test.ts (hunk 10 above)
Comment (esauerbo) [reply packages/react/src/hooks/__tests__/useGetURL.test.ts]: What do you mean by that? This line is clears the `getUrl` mock between tests

File: packages/react-storage/src/components/StorageImage/StorageImage.tsx [hunk 12]
```diff
@@ -1,53 +1,23 @@
-import * as React from 'react';
```
Comment (reesscot) [inline packages/react-storage/src/components/StorageImage/StorageImage.tsx:1]: I'm not sure we can remove this, I think it's needed by React 16? https://legacy.reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html#removing-unused-react-imports

Will let @calebpollman confirm.

File: packages/react/src/hooks/useGetURL.ts (hunk 6 above)
Comment (esauerbo) [reply packages/react/src/hooks/useGetURL.ts]: Yeah I think we can just define it inside of the `useEffect`

**Comment by esauerbo** [comment]
Closing in favor of smaller prs that break this task up.
